File: examples/studio/web/reindexing.js
// Copy/address decisions only. Execution and occurrence lineage belong to Python.
const el=(tag,text,attrs={})=>{const n=document.createElement(tag);if(text!==undefined)n.textContent=text;for(const [k,v] of Object.entries(attrs))n.setAttribute(k,v);return n};
const labeled=(box,text,node)=>{const label=el('label',text);label.append(node);box.append(label);return node};
function options(entries,id){const s=el('select',undefined,{id});for(const [value,text] of entries)s.append(el('option',text,{value}));return s}

export function reindexControls(source,objects){
  const box=el('div');
  const kind=labeled(box,'Operation',options([['tile','Repeat'],['gather','Take an address list'],['concat','Join another object']],'reindex-kind'));
  const axis=labeled(box,'Along',options([...source.axes.map(a=>[a,`Axis ${a}`]),['','Flattened sequence']],'reindex-axis'));
  axis.value=source.axes[0]||'';
  const placement=labeled(box,'Result placement',options([['indices','Index chart'],['unplaced','Unplaced · arrange later']],'reindex-placement'));
  const explanation=el('p',undefined,{class:'help',id:'reindex-meaning'});box.append(explanation);
  const panels={};for(const mode of ['tile','gather','concat']){panels[mode]=el('div');box.append(panels[mode])}
  const candidates=objects.filter(o=>o.status==='ready'&&o.kind!=='incidence');
  const repeatMode=labeled(panels.tile,'Repeat count from',options([['formula','A number or parameter formula'],['object','A single value from an object']],'reindex-repeat-mode'));
  const formulaBox=el('div'),objectBox=el('div');panels.tile.append(formulaBox,objectBox);
  const times=labeled(formulaBox,'Number of repeats',el('input',undefined,{id:'reindex-times',value:'2',inputmode:'text'}));
  const repeatSource=labeled(objectBox,'Count object',options(candidates.filter(o=>o.fields.includes('value')).map(o=>[o.name,o.name]),'reindex-repeat-source'));
  panels.tile.append(el('p','One nonnegative integer is required. An object must contain exactly one value; that dependency is retained when parameters change. Zero repeats gives an empty domain.',{class:'help'}));
  function repeatChoice(){formulaBox.hidden=repeatMode.value!=='formula';objectBox.hidden=repeatMode.value!=='object'}
  repeatMode.onchange=repeatChoice;repeatChoice();
  const addresses=labeled(panels.gather,'Address object',options(candidates.map(o=>[o.name,o.name]),'reindex-addresses'));
  const fields=labeled(panels.gather,'Read addresses from',options([],'reindex-field'));
  const order=labeled(panels.gather,'Order address rows by',options([],'reindex-order'));
  const claim=labeled(panels.gather,'Coverage requirement',options([['copies','Allow repeats and omissions'],['bijective','Every source slot exactly once']],'reindex-claim'));
  panels.gather.append(el('p','Addresses are zero-based source slots, not labels or keys. One address supplies each destination slot. Order must be unique; ties fail. The same address list applies to every fiber along a chosen axis.',{class:'help'}));
  function addressFields(){const obj=candidates.find(o=>o.name===addresses.value);for(const s of [fields,order]){s.replaceChildren();for(const f of obj?.fields||[])s.append(el('option',f,{value:f}))}fields.value=obj?.fields.includes('value')?'value':obj?.fields[0]||'';order.value='index'}
  addresses.onchange=addressFields;addressFields();
  const other=labeled(panels.concat,'Append this object',options(candidates.map(o=>[o.name,o.name]),'reindex-other'));
  panels.concat.append(el('p','All other axis sizes and attribute names must agree. For zero padding, create a zero-valued grid explicitly, then join it. Tuple-only inputs can join other tuple-only inputs.',{class:'help'}));
  function show(){
    for(const [mode,panel] of Object.entries(panels))panel.hidden=mode!==kind.value;
    const axes=axis.value?[...source.axes].reverse():['index'];
    const dimension=source.dimension||Math.max(2,axes.length);while(axes.length<dimension)axes.push('0');
    explanation.textContent=`New occurrences retain captured source links; a repeated address makes distinct copies. ${placement.value==='indices'?`Index chart: (${axes.join(', ')}).`:'No positions are supplied.'} Use a new Result name to keep the source beside it, or its current name to record this step on the source. Earlier consumers retain their inputs.`;
  }
  kind.onchange=axis.onchange=placement.onchange=show;show();
  return {box,read:()=>({source:source.name,kind:kind.value,axis:axis.value||null,placement:placement.value,
    ...(kind.value==='tile'?{times:repeatMode.value==='object'?{object:repeatSource.value}:{formula:times.value}}:kind.value==='concat'?{other:other.value}:{addresses:addresses.value,field:fields.value,order:order.value,bijective:claim.value==='bijective'})})};
}

File: examples/studio/reindexing.py
"""Authoring recipes for copied domains; the core owns addressing and lineage.

No view coordinates become source addresses. Ordering is a checked construction,
not the current order of marks in a rendered table.
"""

from kaleion import F
from kaleion.model import Snapshot


def reindex(args, roots, captured, parse):
    common = {"source", "kind", "axis", "placement"}
    choices = {"tile": {"times"}, "gather": {"addresses", "field", "order", "bijective"},
               "concat": {"other"}}
    kind = args.get("kind")
    if kind not in choices or set(args) != common | choices[kind]:
        raise ValueError("Choose Repeat, Address list or Join and supply its choices")
    source = roots[args["source"]]
    snapshot = captured.results.get(args["source"])
    if not isinstance(snapshot, Snapshot):
        raise ValueError("Choose a ready collection; measure or select an incidence first")
    axis = args["axis"]
    if axis is not None and (not isinstance(axis, str) or axis not in snapshot.axes):
        raise ValueError("Choose a declared logical axis or the flattened sequence")
    if args["placement"] not in ("unplaced", "indices"):
        raise ValueError("Choose Unplaced or Index chart for the result")
    if kind == "tile":
        times = args["times"]
        if isinstance(times, dict) and set(times) == {"object"}:
            # A constructor takes one value, not one keyed read per source item.
            # scalar() checks this again in every parameter case, including empty
            # or multiple measurements. Never freeze the current displayed total.
            driver = captured.results.get(times["object"])
            if not isinstance(driver, Snapshot):
                raise ValueError("Choose a ready object supplying one integer value")
            times = roots[times["object"]].scalar()
        else:
            times = parse(times)
        result = source.tile(times, axis=axis)
    elif kind == "concat":
        other = captured.results.get(args["other"])
        if not isinstance(other, Snapshot):
            raise ValueError("Join another ready collection")
        result = source.concat(roots[args["other"]], axis=axis)
    else:
        address_snapshot = captured.results.get(args["addresses"])
        if not isinstance(address_snapshot, Snapshot):
            raise ValueError("Choose a ready address collection")
        if any(not isinstance(args[k], str) or args[k] not in address_snapshot.context()
               for k in ("field", "order")):
            raise ValueError("Choose an address field and an ordering field from the address object")
        if type(args["bijective"]) is not bool:
            raise ValueError("Declare whether every source slot must occur exactly once")
        addresses = roots[args["addresses"]]
        # Checking a report once is insufficient: this guard survives new cases.
        unique_order = addresses.count(by=F[args["order"]]).where(F.value == 1)
        ordered = addresses.require(unique_order, message="Address order must be unique")
        ordered = ordered.order_by(F[args["order"]]).with_values(F[args["field"]])
        operation = source.permute if args["bijective"] else source.gather
        result = operation(ordered, axis=axis)
    if args["placement"] == "indices":
        # A declared display chart, separate from the source-address decision.
        axes = list(reversed(snapshot.axes)) if axis is not None else []
        coordinates = [F[a] for a in axes] or [F.index]
        dimension = snapshot.positions.shape[1] if snapshot.positions is not None else max(2, len(coordinates))
        if len(coordinates) > dimension:
            raise ValueError("Index chart needs all logical axes; choose Unplaced, then Arrange")
        result = result.arrange(*coordinates, *([0] * (dimension - len(coordinates))))
    return result
